command/agent/monitor/stream_helpers.go

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package monitor
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"strings"
+	"sync"
+	"syscall"
+
+	"github.com/hashicorp/go-msgpack/v2/codec"
+	sframer "github.com/hashicorp/nomad/client/lib/streamframer"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// Stream Helpers
+type StreamReader struct {
+	sync.Mutex
+	framer *sframer.StreamFramer
+	ch     <-chan []byte
+	buf    []byte
+}
+
+func NewStreamReader(ch <-chan []byte, framer *sframer.StreamFramer) *StreamReader {
+	return &StreamReader{
+		ch:     ch,
+		framer: framer,
+	}
+
+}
+
+func (r *StreamReader) Read(p []byte) (n int, err error) {
+
+	select {
+	case data, ok := <-r.ch:
+		if !ok && len(data) == 0 {
+			return 0, io.EOF
+		}
+		r.Lock()
+		r.buf = data
+	default:
+		return 0, nil
+	}
+
+	n = copy(p, r.buf)
+	r.buf = r.buf[n:]
+	r.Unlock()
+	return n, nil
+}
+
+func (r *StreamReader) StreamFixed(ctx context.Context, offset int64, path string, limit int64,
+	eofCancelCh chan error, cancelAfterFirstEof bool) error {
+	defer r.framer.Flush()
+	parseFramerErr := func(err error) error {
+		if err == nil {
+			return nil
+		}
+		errMsg := err.Error()
+
+		if strings.Contains(errMsg, io.ErrClosedPipe.Error()) {
+			// The pipe check is for tests
+			return syscall.EPIPE
+		}
+
+		// The connection was closed by our peer
+		if strings.Contains(errMsg, syscall.EPIPE.Error()) || strings.Contains(errMsg, syscall.ECONNRESET.Error()) {
+			return syscall.EPIPE
+		}
+
+		if strings.Contains(errMsg, "forcibly closed") {
+			return syscall.EPIPE
+		}
+
+		return err
+	}
+	// streamFrameSize is the maximum number of bytes to send in a single frame
+	streamFrameSize := int64(512)
```

Comment:
This is hard-coded but the callers set the size of the monitor (512) and framer (1024!?) elsewhere. Is there a reason this doesn't use the framer size?

Reply:
Thanks for this one! I added this as a parameter to NewStreamReader so that we could pass the same value to the StreamReader and the StreamFramer